### ISSUE TITLE
[Feature Request] Invalid action masking

### ISSUE THREAD

**Issue by kronion** [Stable-Baselines-Team/stable-baselines3-contrib#16]
### :rocket: Feature

Add support for masking out invalid actions when making predictions. This would allow models to converge much faster for environments where many actions may be illegal depending on the state.

This feature could be implemented as a wrapper around a gym environment that adds a method to return the mask. The stable baselines algorithms would check for the wrapper and use the mask if available. The mask is a boolean tensor in the shape of the action space, and it replaces the logits for invalid actions with very large negative values in the underlying probability distribution.

[Here is an MVP](https://github.com/kronion/stable-baselines3/compare/master...kronion:feat/action_masking).

### Motivation

In many environments, there may be portions of the action space that are invalid to select in a given state. Without a way to avoid sampling these actions, training becomes less efficient. Models have to waste time exploring the invalid portions of the space, which may become prohibitively expensive for large action spaces. See [this paper](https://arxiv.org/pdf/2006.14171.pdf) for more details.

### Alternatives

To the best of my knowledge, the only alternative is to accept that invalid actions may be selected, and to try to discourage it by penalizing the choice with a large negative reward. This is just the status quo. Note that action masking would be optional, and the status quo would stay the default.

### Additional context

I ran into this problem in practice when building models for board games. To work around it, I implemented an MVP of this feature in a fork. I'd be happy to make a PR. The branch is [here](https://github.com/kronion/stable-baselines3/tree/feat/action_masking).

Also related: hill-a/stable-baselines#453

Original issue here: https://github.com/DLR-RM/stable-baselines3/issues/269 

### Checklist

- [x] I have checked that there is no similar issue in the repo (required)

**Comment by H-Park** [comment]
Does this masking approach match the pre print exploration of action masking? Did you verify the policies don't diverge and all that goodness?

**Comment by kronion** [comment]
> Does this masking approach match the pre print exploration of action masking?

That was my intention, yes. I implemented strategy 2 from the paper. Masks don't have to be provided during testing, either, so you can run with "masking removed" as well.

> Did you verify the policies don't diverge and all that goodness?

I have not verified this, but I'd be happy to do so. I've only confirmed that my models converge to similar expected episode rewards with and without masking, and that convergence is faster with masking.

**Comment by Miffyli** [comment]
Dropping by to mention that we will get back to this after holidays with refreshed minds :)

**Comment by Miffyli** [comment]
After a preliminary look over the fork the overall structure seems good, especially the idea of using `apply_masking` function in distribution (makes it clear which distributions support masking and is easy to call). My main gripe is with `action_space_fn` which is bit hacky. The idea of using dictionary observations to include action-space mask sounds most intuitive, which should be doable after [support for dict-obs is merged](https://github.com/DLR-RM/stable-baselines3/pull/243). 

> I have not verified this, but I'd be happy to do so. I've only confirmed that my models converge to similar expected episode rewards with and without masking, and that convergence is faster with masking.

For this repo we are interested in correctness of implementations, so matching results of a reference code or reference paper is more important than "getting good results" (users of the code can be judge of that).

@vwxyzjn Pinging you in case you have some insights/comments to provide :)

**Comment by araffin** [comment]
I will let @Miffyli handle the rest of the discussion but I would favor implementing it first as a duplicated algorithm, for instance `PPO_MASK` instead of trying to enable that experimental feature on all possible algorithms.

**Comment by vwxyzjn** [comment]
Hey @Miffyli thanks for pinning me :). @kronion Out of curiosity, what game environment did you test the invalid action masking on?


>  My main gripe is with action_space_fn which is bit hacky. The idea of using dictionary observations to include action-space mask sounds most intuitive, which should be doable after support for dict-obs is merged.

A `env.valid_actions()` might be quite necessary. The problem with using a dictionary that contains action mask is that the action mask might be conditional on some part of the actions (this phrasing is probably very confusing, I apologize; let me explain with an example).

In gym-microrts, we have a `MultiDiscrete([100   6   4   4   4   4   7 100])` action space. So the first number (out of 100) of a sampled action might look like `[2,1,3,0,0,0,0,0]`, which means the unit at location 2 is executing the move action (of action type 1) and in the west direction (of move action parameter 3). Note that the action type and parameter is `conditioned` upon the unit selected to do actions.

So in practice, we use two functions to get different part of the invalid action mask to generate an action:

```python
# https://github.com/vwxyzjn/gym-microrts/blob/146e6b9255a9799abe4b85889264ba6e6a791b58/experiments/ppo.py#L231-L244
# 1. select source unit based on source unit mask
source_unit_mask = env.get_unit_location_mask()
multi_categoricals = [CategoricalMasked(logits=split_logits[0], masks=source_unit_mask)]
action_components = [multi_categoricals[0].sample()]
# 2. select action type and parameter section based on the
#    source-unit mask of action type and parameters
source_unit_action_mask = env.get_unit_action_mask(unit=action_components[0][i])
...
```

Alternatively, it is possible to include the action type and parameters mask for every possible unit in the game, but it might be quite memory intensive (in a 10x10 map there is 100 locations)

**Comment by kronion** [comment]
Apologies for the slow reply, @vwxyzjn. Initially I was experimenting with a game environment I wrote myself. [I've published it here](https://github.com/kronion/gym-love-letter) in case you'd like to take a look at it, but I don't expect it to be easy for others to run it at the moment.

More recently I've actually been focused on testing against gym-microrts, with the hopes of confirming that I get similar results to the reference paper (per Miffyli's point above). I'm basically trying to adapt [the experiments in your repo](https://github.com/vwxyzjn/invalid-action-masking) to use SB3 instead, but I haven't had enough time to complete the work.

Regarding the `env.valid_actions()` approach, I agree that `action_mask_fn` is hacky. I don't like how many function signatures it changes. I think it might be better if it's the caller's responsibility to pass the algorithm an environment that's already wrapped with an `ActionMasker` if masking is desired. That way the signatures don't have to change because it wouldn't be the algorithm's responsibility to create the wrapper.

**Comment by H-Park** [comment]
I just want to chime in one more time and thank @kronion for picking up the torch on this and seeing it through. I just verified your branch and it works extremely well and is very nice to use from a programming perspective.

**Comment by araffin** [comment]
I only had a quick glance at it, but look like a good (visual) resource too: https://boring-guy.sh/posts/masking-rl/

**Comment by qorrect** [comment]
I'd like to use it could you point me to where to begin to implement this in my own environment ?  Thanks!

**Comment by H-Park** [comment]
```
def mask_function(env):
    return env.action_mask
model = PPO(MlpPolicy, env, verbose=1, action_mask_fn=mask_function)
```

The big "gotcha" is that **action_mask_fn** is a callback to your environment **_instance_**'s action mask method. This does one key thing: allows for vectorized action masking implicitly, since each environment takes care of masking via a callback. This is a major improvement from the action masking PR for stable baselines 2, which didn't support stacked environments. 

Massive props to @kronion for that clean API design.

**Comment by qorrect** [comment]
This is working wonderfully for me, thanks guys appreciate you!

**Comment by qorrect** [comment]
What would the mask look like for a MultiDiscrete action space of (64,64) ?  And is there a better place to ask this ?  Thanks!

**Comment by H-Park** [comment]
Depends on the nature of the 64, 64 space. Are the actions linked for masking? For instance, in chess, you select what piece you want to move, then where to move it. These decisions impact each other. 

If this is the case, you need a masking structure of size (64) to mask your first action, then (64,64) for your second action, where the first index of the second mask is the action selected from the first action. 

If this is not the case, then a simple (2,64) would work. 

I'm not sure there is a "best spot" to ask usage questions for an unmerged PR. Maybe an entry should be added to the docs to showcase this feature and the possible use cases - action masking for invalid actions, machine teaching, etc

**Comment by qorrect** [comment]
This is for checkers so the first option ( linked masking ).

So when you say two actions, one to select the existing piece and another action to move it to the new index, is that two step() calls ? 

Truly appreciative of your responses @H-Park thank you.

**Comment by H-Park** [comment]
It really depends on your model set up, and I don't think there is ANY literature regarding a hierarchical masking policy, given that action masking in general is so new. 

I would have a model with two outputs: (12, 64): (piece, square), both one-hot. The masking arrays would be 12, and (12, 64). For one step, you first decide which piece to move from the 12, according to the mask (remaining pieces on the board), then decide where to move it from the 64 possible squares, masked according to the piece you're trying to move and the possible legal moves it can take. 

Of course this is a bad example because the action space can be heavily reduced: checker pieces can only move 2 ways normally, 4 if kinged, but you get the idea. 

This is how I masked my StarCraft II [agent](https://github.com/H-Park/starcraft2ai/blob/8857901b3658b366dd2aeaebefe50cec3ed67ab9/A3C/agents/a3c_agent.py#L141-L143)

**Comment by kronion** [comment]
I just started exploring adding MultiDiscrete support in #25 and promptly ran into this same question. Sharing the two options I've considered so far:
1. Require each discrete component of the MultiDiscrete action space to be independent of all the rest. Each discrete component would have its own categorical distribution with its own mask. This makes the implementation straightforward on SB3's end, but it means that the writer of the gym environment is responsible for grouping _dependent_ actions together as a single "joint" action that expresses the union. For example, rather than representing the checkers action as (12, 64), it would be (12 * 64). Note that this is not the same as representing the action space as one big categorical choice -- only the mutually dependent actions need to be merged. Gym environments that do not follow this convention could be wrapped to translate the action shape back and forth between the joint and independent representations.
2. Let environments continue to represent mutually dependent actions as independent pieces of a MultiDiscrete action (ie the status quo), but add a way to pass additional context to the SB3 agent so that these actions to be represented jointly under the hood. This is basically the same as the first suggestion, but it requires changes both by SB3 and by environment writers.

I prefer option 1 because it's a simpler interface, all outside of SB3 agent code.

I hadn't considered selecting the mutually dependent actions one at a time in a hierarchical manner. It's not clear to me how this would be implemented, but it would presumably change the way the agent interacts with the environment, both during training and at runtime. The shape of the action space would depend on the step the agent is on, which could be tricky to handle. It does seem like an interesting avenue for further exploration, though!

**Comment by Miffyli** [comment]
> I prefer option 1 because it's a simpler interface, all outside of SB3 agent code.

This sounds preferable for the reasons you stated. Only if it becomes impractical for some reason (e.g. too big action space), then something else can be thought up (e.g. like the second option).
